#' Plot processed data figures
#'
#' Reproduces processed data figure panels
#'
#' @details This function reproduces the following figure panels: \itemize{\item
#'   \strong{Figure 1G}: UMAP projection of cells cultured in IL22+ medium,
#'   colored by cell-type cluster \item \strong{Figure 1H}: Dot plot depicting
#'   selected marker gene expression aggregated per cell-type cluster \item
#'   \strong{Figure 2C}: Bar plot showing the difference in cell-type cluster
#'   proportions between cells cultured in IL22+ and IL22- media \item
#'   \strong{Figure S2A}: UMAP plots depicting normalized gene expression of
#'   cell-type markers in IL22+ cultured cells \item \strong{Figure S2B}: UMAP
#'   plots depicting normalized gene expression of WNT ligands in IL22+ cultured
#'   cells \item \strong{Figure S6A}: Integrated UMAP projection of all cells
#'   colored by culture condition (IL22+/IL22-) \item \strong{Figure S6B}:
#'   Alluvial plot highlighting the proportions of each cell-type cluster per
#'   culture condition (IL22+/IL22-) \item \strong{Figure S7A}: UMAP plots
#'   depicting normalized gene expression of IL22 receptor genes in cells
#'   cultured in either IL22+ or IL22- medium}
#'
#' @param object The output Seurat object generated by
#'   \code{\link{process_data}}
#' @param figure Character ID of the figure to be reproduced. Accepted values are "1G",
#'   "1H", "2C", "S2A", "S2B", "S6A", "S6B" and "S7A" (See Details for
#'   description of figure panels)
#' @param save Save the specified plot as a pdf? (default = FALSE)
#' @param save.dir Path to the directory where the output plot should be saved
#'   (default = ".")
#'
#' @return The plot specified in the "figure" argument
#' @export
#'
#' @examples
#' \dontrun{
#' # Plot all figures, save all plots as pdf files in the current working directory
#' plot_processed(object = final.output, figure = "1G", save = T, save.dir = ".")
#' plot_processed(object = final.output, figure = "1H", save = T, save.dir = ".")
#' plot_processed(object = final.output, figure = "2C", save = T, save.dir = ".")
#' plot_processed(object = final.output, figure = "S2A", save = T, save.dir = ".")
#' plot_processed(object = final.output, figure = "S2B", save = T, save.dir = ".")
#' plot_processed(object = final.output, figure = "S6A", save = T, save.dir = ".")
#' plot_processed(object = final.output, figure = "S6B", save = T, save.dir = ".")
#' plot_processed(object = final.output, figure = "S7A", save = T, save.dir = ".")}
plot_processed <- function(object=NULL, figure=NA, save=FALSE, save.dir="."){

  # Check inputs
  accept.figs <- c("1G","1H","2C","S2A","S2B","S6A","S6B","S7A")

  if (is.na(figure)){
    stop("Please specify a figure to plot")
  }

  if (!is.element(figure,accept.figs)){
    stop("Invalid figure ID, please specify one of the following: ", paste(accept.figs,collapse = ", "))
  }

  # General variables

  clust.order <- c("ISC","TA1","TA2","Early EC","EC","Secretory progenitor","Goblet","Paneth","EEC","Tuft")
  clust.cols <- c("#8b8c89","#FAB198", "#BF212F","#33AD75","#006F3C","#AACEDC","#254B95","#FFB60A","#CB157C","#76018D")


  # Plot figure 1G
  if (figure == "1G"){

    data <- Seurat::FetchData(object, vars = c("UMAP_1","UMAP_2","Final.IDS"),
                              cells = Seurat::WhichCells(object, expression = condition == "IL22+"))

    # Chose random cells to label (per cluster)
    set.seed(27)
    data.sub <- data %>% dplyr::group_by(Final.IDS) %>% dplyr::sample_n(1)

    data$id <- factor(data$Final.IDS, levels = clust.order)

    plot <- ggplot2::ggplot(data, ggplot2::aes(x=UMAP_1, y=UMAP_2, color=Final.IDS)) +
      ggplot2::geom_point(shape=20, size=1.5) +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
            panel.grid.minor = ggplot2::element_blank(),
            panel.background = ggplot2::element_rect(colour = "black", size=1),
            axis.ticks.length = ggplot2::unit(.25, "cm"),
            legend.position = "none",
            axis.title = ggplot2::element_text(size=14, face="bold"),
            axis.text = ggplot2::element_text(size=12)) +
      ggplot2::scale_color_manual(values = clust.cols,
                         breaks = clust.order) +
      ggrepel::geom_label_repel(data=data.sub, ggplot2::aes(x=UMAP_1, y=UMAP_2,label=Final.IDS), size=6, force = 100)

    # Save?
    if (save == TRUE){
      ggplot2::ggsave(plot=plot,paste0(save.dir,"/Figure_1G.pdf"), width = 8, height = 8)
    }

    # Draw plot
    return(plot)
  }

  # Plot figure 1H
  if (figure == "1H"){
    # Genes to plot
    highlights <- c("SMOC2","OLFM4","CCND1","CCDC85B","MCM6","MKI67",
                    "FABP1","PCK1","KRT20","IL32","DLL1","HES6","muc2-mneongreen","REG4","PLA2G2A","DEFA6",
                    "NEUROD1","CHGA","AVIL","ALOX5")

    # Take advantage of Seurat::DotPlot to return useful data
    data <- Seurat::DotPlot(subset(object, cells = Seurat::WhichCells(object, expression = condition == "IL22+")),
                    features = highlights, assay = "SCT", group.by = "Final.IDS")
    data <- data$data
    data$id <- factor(data$id, levels = rev(clust.order))
    data$features.plot <- factor(data$features.plot, levels = (highlights))

    plot <- suppressWarnings(ggplot2::ggplot(data, ggplot2::aes(y = id, x = features.plot, size = pct.exp, fill = avg.exp.scaled)) +
      ggplot2::geom_point(shape=21) +
      ggplot2::theme_bw() +
      ggplot2::scale_size_area(max_size = 10) +
      ggplot2::scale_fill_gradient2(low="blue3",high="red3",mid="white",limits=c(-2.5,2.5)) +
      ggplot2::labs(fill="Relative expression", size = "Percent expressed") +
      ggplot2::xlab("Cluster") +
      ggplot2::ylab("Gene") +
      ggplot2::theme(axis.text.y = ggplot2::element_text(size=12,color=rev(clust.cols)),
            axis.text.x = ggplot2::element_text(size=12,angle = 45, hjust=1),
            axis.title = ggplot2::element_text(size=0),
            legend.title=ggplot2::element_text(size=14),
            legend.text = ggplot2::element_text(size=12),
            legend.position = "bottom",
            panel.background = ggplot2::element_rect(colour = "black", size=1)) +
      ggplot2::scale_x_discrete(labels=c("SMOC2","OLFM4","CCND1","CCDC85B","MCM6","MKI67",
                                "FABP1","PCK1","KRT20","IL32","DLL1","HES6","MUC2","REG4","PLA2G2A","DEFA6",
                                "NEUROD1","CHGA","AVIL","ALOX5")) +
      ggplot2::guides(size=ggplot2::guide_legend(ncol=2, title.position = "top"),
             fill=ggplot2::guide_colorbar(title.position = "top", label.position = "top")))



    if (save == TRUE){
      ggplot2::ggsave(plot=plot,paste0(save.dir,"/Figure_1H.pdf"), width = 8, height = 8)
    }

    # Draw plot
    return(plot)

  }

  # Plot figure 2C
  if (figure == "2C"){

    data <- prop.table(table(object$Final.IDS, object$condition),2)
    fc <-log2(data[,"IL22+"]) - log2(data[,"IL22-"])

    data <- cbind.data.frame("FC"=fc, "Cluster"=names(fc))

    data$Cluster <- factor(data$Cluster, levels=rev(clust.order))

     plot <- ggplot2::ggplot(data, ggplot2::aes(x=FC, y=Cluster,fill=Cluster)) +
      ggplot2::geom_bar(stat="identity") +
      ggplot2::theme_bw() +
      ggplot2::scale_fill_manual(values = clust.cols,
                        breaks = clust.order) +
      ggplot2::theme(panel.background = ggplot2::element_rect(colour = "black", size=1),
            legend.position = "none",
            axis.title = ggplot2::element_text(size=14, face="bold"),
            axis.text = ggplot2::element_text(size=12),
            panel.grid.major.y = ggplot2::element_blank(),
            panel.grid.minor.x = ggplot2::element_blank()) +
      ggplot2::xlab("Frequency FC (Log2)") +
      ggplot2::ylab("")


    if (save == TRUE){
      ggplot2::ggsave(plot=plot,paste0(save.dir,"/Figure_2C.pdf"), width = 5, height = 6)
    }

     # Draw plot
     return(plot)
  }

  # Plot Figure S2A
  if (figure == "S2A"){

    markers <- c("SMOC2","MKI67","FABP1","KRT20","HES6","muc2-mneongreen","DEFA6","CHGA","AVIL")

    labels <- c("SMOC2 (Stem cells)","MKI67 (Proliferating cells)","FABP1 (Early enterocytes)",
                "KRT20 (Enterocytes)","HES6 (Secretory progenitors)","MUC2 (Goblet cells)",
                "DEFA6 (Paneth cells)","CHGA (Enteroendocrine cells)","AVIL (Tuft cells)")

    plot.list <- list()

    for (i in 1:length(markers)){
      gene <- markers[i]

      label <- labels[i]

      data <- Seurat::FetchData(object, vars = c("UMAP_1","UMAP_2",gene),
                                cells = Seurat::WhichCells(object, expression = condition == "IL22+"))

      pl <- ggplot2::ggplot(data %>% dplyr::arrange(!!ggplot2::sym(gene)),
                            ggplot2::aes_string(x="UMAP_1",y="UMAP_2",color=paste0("`",gene,"`"))) +
        ggplot2::geom_point(shape=20, size=1) +
        viridis::scale_color_viridis(option = "D", direction=-1) +
        ggplot2::theme_classic() +
        ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
              panel.grid.minor = ggplot2::element_blank(),
              axis.ticks.length = ggplot2::unit(.25, "cm"),
              axis.title = ggplot2::element_blank(),
              axis.text = ggplot2::element_text(size=12),
              plot.title = ggplot2::element_text(size=16,face = "bold"),
              legend.title = ggplot2::element_blank()) +
        ggplot2::labs(title = label)

      plot.list[[gene]] <- pl

    }
    # Plot in grid
    plot <- cowplot::plot_grid(plotlist =  plot.list)


    if (save == TRUE){
      ggplot2::ggsave(plot=plot,paste0(save.dir,"/Figure_S2A.pdf"), width = 12, height = 12)
    }

    return(plot)
  }

  # Plot Figure S2B
  if (figure == "S2B"){

    markers <- c("WNT2B","WNT3","WNT4","WNT5B","WNT7B","WNT10A","WNT11")

    plot.list <- list()

    for (i in 1:length(markers)){
      gene <- markers[i]

      data <- Seurat::FetchData(object, vars = c("UMAP_1","UMAP_2",gene),
                                cells = Seurat::WhichCells(object, expression = condition == "IL22+"))

      pl <- ggplot2::ggplot(data %>% dplyr::arrange(!!ggplot2::sym(gene)),
                            ggplot2::aes_string(x="UMAP_1",y="UMAP_2",color=paste0("`",gene,"`"))) +
        ggplot2::geom_point(shape=20, size=1) +
        viridis::scale_color_viridis(option = "D", direction=-1) +
        ggplot2::theme_classic() +
        ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
              panel.grid.minor = ggplot2::element_blank(),
              axis.ticks.length = ggplot2::unit(.25, "cm"),
              axis.title = ggplot2::element_blank(),
              axis.text = ggplot2::element_text(size=12),
              plot.title = ggplot2::element_text(size=16,face = "bold"),
              legend.title = ggplot2::element_blank()) +
        ggplot2::labs(title = gene)

      plot.list[[gene]] <- pl

    }
    plot <- cowplot::plot_grid(plotlist =  plot.list, nrow = 2)

    if (save == TRUE){
      ggplot2::ggsave(plot=plot,paste0(save.dir,"/Figure_S2B.pdf"), width = 12, height = 5)
    }
    return(plot)

  }

  # Plot Figure S6A
  if (figure == "S6A"){

    data <- Seurat::FetchData(object, vars = c("UMAP_1","UMAP_2","condition"))

    # Shuffle cells
    set.seed(27)
    data <- data %>% dplyr::sample_n(nrow(data))

    plot <- ggplot2::ggplot(data, ggplot2::aes(x=UMAP_1, y=UMAP_2, color=condition)) +
      ggplot2::geom_point(shape=20, size=1.5) +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
            panel.grid.minor = ggplot2::element_blank(),
            panel.background = ggplot2::element_rect(colour = "black", size=1),
            axis.ticks.length = ggplot2::unit(.25, "cm"),
            axis.title = ggplot2::element_text(size=14, face="bold"),
            axis.text = ggplot2::element_text(size=12),
            legend.title=ggplot2::element_text(size=14),
            legend.text = ggplot2::element_text(size=12),
            legend.position = c(.9, .1)) +
      ggplot2::scale_color_manual(values = c("#00547A","#999999"),
                         breaks = c("IL22+","IL22-")) +
      ggplot2::labs(color="Condition") +
      ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size = 6)))



    if (save == TRUE){
      ggplot2::ggsave(plot=plot,paste0(save.dir,"/Figure_S6A.pdf"), width = 8, height = 8)
    }
    # Draw plot
    return(plot)

  }

  # Plot Figure S6B (Apdapted from code written by Lin)
  if (figure == "S6B"){
    cell_ct <- object@meta.data %>%
      dplyr::group_by(Final.IDS,condition) %>%
      dplyr::count() %>%
      dplyr::group_by(condition) %>%
      dplyr::mutate(percent=100*n/sum(n)) %>%
      dplyr::ungroup() %>%
      as.data.frame
    cell_ct$Final.IDS <- factor(cell_ct$Final.IDS, levels = clust.order)

    plot <- ggplot2::ggplot(cell_ct, ggplot2::aes(x=condition, y= percent, stratum=Final.IDS, alluvium = Final.IDS,
                                                         fill = Final.IDS, label = Final.IDS)) +
      ggplot2::scale_fill_manual(values = clust.cols,
                                 breaks = clust.order) +
      ggplot2::scale_x_discrete(limits = rev(levels(cell_ct$condition)))+
      ggalluvial::geom_flow(stat = "alluvium", lode.guidance = "frontback",
                color = "black", curve_type = "cubic", alpha = 1/3, reverse = FALSE) +
      ggalluvial::geom_stratum(reverse = FALSE) +
      ggplot2::coord_flip()+
      ggplot2::theme(legend.position = "top", panel.background = ggplot2::element_blank()) +
      ggplot2::ggtitle("Percentage of cell types per condition")


    if (save == TRUE){
      ggplot2::ggsave(plot=plot, file = paste0(save.dir,"/Figure_S6B.pdf"), width = 10, height = 4)
    }
    # Draw plot
    return(plot)

  }

  # Plot Figure S7A
  if (figure == "S7A"){

    markers <- c("IL22RA1","IL10RB")
    cond <- c("IL22+","IL22-")

    plot.list.IL22RA1 <- list()
    for (i in 1:length(cond)){
      data <- Seurat::FetchData(object, vars = c("UMAP_1","UMAP_2","IL22RA1"),
                                cells = Seurat::WhichCells(object, expression = condition == cond[i]))

      pl <- ggplot2::ggplot(data %>% dplyr::arrange(IL22RA1),
                            ggplot2::aes_string(x="UMAP_1",y="UMAP_2",color="IL22RA1")) +
        ggplot2::geom_point(shape=20, size=1) +
        viridis::scale_color_viridis(option = "D", direction=-1) +
        ggplot2::theme_classic() +
        ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
                       panel.grid.minor = ggplot2::element_blank(),
                       axis.ticks.length = ggplot2::unit(.25, "cm"),
                       axis.title = ggplot2::element_blank(),
                       axis.text = ggplot2::element_text(size=12),
                       plot.title = ggplot2::element_text(size=16,face = "bold"),
                       legend.title = ggplot2::element_blank()) +
        ggplot2::labs(title = cond[i])

      plot.list.IL22RA1[[cond[i]]] <- pl
    }

    plot.list.IL10RB <- list()
    for (i in 1:length(cond)){
      data <- Seurat::FetchData(object, vars = c("UMAP_1","UMAP_2","IL10RB"),
                                cells = Seurat::WhichCells(object, expression = condition == cond[i]))

      pl <- ggplot2::ggplot(data %>% dplyr::arrange(IL10RB),
                            ggplot2::aes_string(x="UMAP_1",y="UMAP_2",color="IL10RB")) +
        ggplot2::geom_point(shape=20, size=1) +
        viridis::scale_color_viridis(option = "D", direction=-1) +
        ggplot2::theme_classic() +
        ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
                       panel.grid.minor = ggplot2::element_blank(),
                       axis.ticks.length = ggplot2::unit(.25, "cm"),
                       axis.title = ggplot2::element_blank(),
                       axis.text = ggplot2::element_text(size=12),
                       plot.title = ggplot2::element_text(size=16,face = "bold"),
                       legend.title = ggplot2::element_blank()) +
        ggplot2::labs(title = cond[i])

      plot.list.IL10RB[[cond[i]]] <- pl
    }

    plot.IL22RA1 <- cowplot::plot_grid(plotlist =  plot.list.IL22RA1, nrow = 1)

    # Code adapted from https://wilkelab.org/cowplot/articles/plot_grid.html
    title <- cowplot::ggdraw() +
      cowplot::draw_label(
        "IL22RA1",
        fontface = 'bold',
        x = 0,
        hjust = 0,
        size = 20
      ) +
      ggplot2::theme(
        plot.margin = ggplot2::margin(0, 0, 0, 7)
      )
    plot.IL22RA1.fin <- cowplot::plot_grid(
      title, plot.IL22RA1,
      ncol = 1,
      rel_heights = c(0.1, 1)
    )

    plot.IL10RB <- cowplot::plot_grid(plotlist =  plot.list.IL10RB, nrow = 1)

    # Code adapted from https://wilkelab.org/cowplot/articles/plot_grid.html
    title <- cowplot::ggdraw() +
      cowplot::draw_label(
        "IL10RB",
        fontface = 'bold',
        x = 0,
        hjust = 0,
        size = 20
      ) +
      ggplot2::theme(
        plot.margin = ggplot2::margin(0, 0, 0, 7)
      )
    plot.IL10RB.fin <- cowplot::plot_grid(
      title, plot.IL10RB,
      ncol = 1,
      rel_heights = c(0.1, 1)
    )

    plot <- cowplot::plot_grid(plotlist = list(plot.IL22RA1.fin,plot.IL10RB.fin), nrow=1)


    if (save == TRUE){
      ggplot2::ggsave(plot=plot, file = paste0(save.dir,"/Figure_S7A.pdf"), width = 10, height = 4)
    }
    #Draw final plot
    return(plot)
  }
}


#' Plot IL22 response figures
#'
#' Reproduces figure panels related to the IL22 response
#'
#' @details This function reproduces the following figure panels: \itemize{\item
#'   \strong{Figure 2D}: Differential gene expression (MA) plot, IL22+ vs IL22-
#'   \item \strong{Figure S7B}: Violin plots depicting the IL22-response module
#'   score between cells cultured in IL22+ and IL22- media (overall and per
#'   cluster) \item \strong{Figure S7D}: Violin plots depicting normalized gene
#'   expression of REG1A, REG1B and DMBT1 between cells cultured in IL22+ and
#'   IL22= media (per cell-type cluster)}
#'
#' @param object The output Seurat object generated by
#'   \code{\link{il22_response}}
#' @param dge.output Path to the DGE file generated by
#'   \code{\link{il22_response}} ("DE_IL22.txt")
#' @param figure ID of the figure to be reproduced. Accepted values are "2D",
#'   "S7B and "S7D" (See Details for description of figure panels)
#' @param save Save the specified plot as a pdf? (default = FALSE)
#' @param save.dir Path to the directory where the output plot should be saved
#'   (default = ".")
#'
#' @return The plot specified in the "figure" argument
#' @export
#'
#' @examples
#' \dontrun{
#' # Make each plot and save as pdf
#' plot_il22(object = il22.output.obj, figure = "2D", dge.output = "DE_IL22.txt", save = T, save.dir = ".")
#' plot_il22(object = il22.output.obj, figure = "S7B",  save = T, save.dir = ".")
#' plot_il22(object = il22.output.obj, figure = "S7D",  save = T, save.dir = ".")}
plot_il22 <- function(object=NULL, dge.output=NA, figure=NA, save=FALSE, save.dir="."){

  # Check inputs
  accept.figs <- c("2D","S7B","S7D")

  if (is.na(figure)){
    stop("Please specify a figure to plot")
  }

  if (!is.element(figure,accept.figs)){
    stop("Invalid figure ID, please specify one of the following: ", paste(accept.figs,collapse = ", "))
  }

  # General variables

  clust.order <- c("ISC","TA1","TA2","Early EC","EC","Secretory progenitor","Goblet","Paneth","EEC","Tuft")
  clust.cols <- c("#8b8c89","#FAB198", "#BF212F","#33AD75","#006F3C","#AACEDC","#254B95","#FFB60A","#CB157C","#76018D")

  # Plot figure 2D

  if (figure == "2D"){

    # Check for diff expression table
    if (is.na(dge.output)){
      stop("Please specify the differential expression file (output from il22_response())")
    }

  data <- read.table(dge.output)

  genes.highlight <- c("REG1A","DMBT1","REG1B","REG3A","OLFM4","REG3G","CD24","STAT3","XBP1","MMP7","PCK1","SI","PHGR1",
                       "REG4")
  highlight <- data$gene
  highlight[!data$gene %in% genes.highlight] <- ""

  data <- cbind.data.frame(data, "label" = highlight)

  plot <- ggplot2::ggplot(data, ggplot2::aes(x=avg_expression, y=avg_log2FC, label=label)) +
    ggplot2::geom_point(data = subset(data, abs(avg_log2FC) < 1) ,shape=20, size=1, color="grey") +
    ggplot2::geom_point(data = subset(data, abs(avg_log2FC) >= 1), shape=20, size=2, color="black") +
    ggplot2::geom_hline(ggplot2::aes(yintercept=1), linetype=2) +
    ggplot2::geom_hline(ggplot2::aes(yintercept=-1), linetype=2) +
    ggplot2::theme_bw() +
    ggplot2::theme(panel.background = ggplot2::element_rect(colour = "black", size=1),
          axis.ticks.length = ggplot2::unit(.25, "cm"),
          axis.title = ggplot2::element_text(size=14, face="bold"),
          axis.text = ggplot2::element_text(size=12),
          panel.grid.minor = ggplot2::element_blank()) +
    ggplot2::labs(x="Mean expression", y=bquote("Average Log"["2"]~"(FC) - IL22+ vs IL22-")) +
    ggrepel::geom_text_repel(max.overlaps = Inf, force=15)



  if (save == TRUE){
    ggplot2::ggsave(plot=plot, file = paste0(save.dir,"/Figure_2D.pdf"), width = 8, height = 6)
  }
  # Draw plot
  return(plot)

  }

  # Plot Figure S7B
  if (figure == "S7B"){
    data <- Seurat::FetchData(object, c("IL22_Module","condition","Final.IDS"))

    data$condition <- factor(data$condition, levels = c("IL22+" , "IL22-"))
    data$Final.IDS <- factor(data$Final.IDS, levels = clust.order)

    # Violin plot per condition
    plot1 <- ggplot2::ggplot(data, ggplot2::aes(x=condition,y=IL22_Module,fill=condition)) +
      ggplot2::geom_violin(scale='width') +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major.x = ggplot2::element_blank(),
            panel.grid.minor.x = ggplot2::element_blank(),
            axis.title.x = ggplot2::element_text(size=14),
            axis.title.y = ggplot2::element_text(size=14),
            axis.text = ggplot2::element_text(size=12),
            legend.title= ggplot2::element_text(size=14),
            legend.text = ggplot2::element_text(size=12),
            legend.position = "none") +
      ggplot2::scale_fill_manual(values = c("#999999","#00547A"),
                        breaks = c("IL22+",  "IL22-")) +
      ggplot2::ylab("IL22 Signature") +
      ggplot2::xlab("") +
      ggplot2::labs(fill = "Condition")

    # Violin plot per cluster per condition
    plot2 <- ggplot2::ggplot(data, ggplot2::aes(x=Final.IDS,y=IL22_Module,fill=condition)) +
      ggplot2::geom_violin(scale='width') +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major.x = ggplot2::element_blank(),
            panel.grid.minor.x = ggplot2::element_blank(),
            axis.text.x = ggplot2::element_text(angle = 45, hjust=1,
                                       color = clust.cols),
            axis.title.x = ggplot2::element_text(size=0),
            axis.title.y = ggplot2::element_text(size=14),
            axis.text = ggplot2::element_text(size=12),
            legend.title= ggplot2::element_text(size=14),
            legend.text = ggplot2::element_text(size=12),
            legend.position = "none") +
      ggplot2::scale_fill_manual(values = c("#999999","#00547A"),
                        breaks = c("IL22+",  "IL22-")) +
      ggplot2::ylab("IL22 Signature") +
      ggplot2::labs(fill = "Condition")

    if (save == TRUE){
      ggplot2::ggsave(plot=plot, file = paste0(save.dir,"/Figure_S7B.pdf"), width = 12, height = 6)
    }
    return(suppressWarnings(cowplot::plot_grid(plotlist = list(plot1,plot2), align = "h", rel_widths = c(1,4))))
  }

  # Plot Figure S7D
  if (figure == "S7D"){

    ### The following code was written by Lin Lin
    condition.cols <- setNames(c("#999999","#00547A"),
                               c("IL22+",
                                 "IL22-"))

    object$condition <- factor(object$condition, levels = c("IL22+","IL22-"))
    object$Final.IDS <- factor(object$Final.IDS, levels = c("ISC", "TA1", "TA2","Early EC","EC","Secretory progenitor", "Goblet","Paneth", "EEC", "Tuft"))

    modify_vlnplot<- function(obj,
                              feature,
                              pt.size = 0,
                              ...) {
      p<- Seurat::VlnPlot(obj, features = feature, pt.size = pt.size, ... )  +
        ggplot2::xlab("") + ggplot2::ylab(feature) + ggplot2::ggtitle("") +
        ggplot2::theme(legend.position = "none",
              axis.text.x = ggplot2::element_blank(),
              axis.title.y = ggplot2::element_text(size = ggplot2::rel(1), angle = 0),
              axis.text.y = ggplot2::element_text(size = ggplot2::rel(1)),
              plot.margin = ggplot2::unit(c(-0.75, 0, -0.75, 0), "cm"))
      return(suppressWarnings(p))
    }

    extract_max<- function(p){
      ymax<- max(ggplot2::ggplot_build(p)$layout$panel_scales_y[[1]]$range$range)
      return(ceiling(ymax))
    }

    # StackedVlnPlot main function
    StackedVlnPlot<- function(obj, features,
                              pt.size = 0,
                              plot.margin = unit(c(-0.75, -0.75, -0.75,-0.75), "cm"),
                              ...) {

      plot_list<- purrr::map(features, function(x) modify_vlnplot(obj = object,feature = x, ...))

      # Add back x-axis title to bottom plot. patchwork is going to support this?
      plot_list[[length(plot_list)]]<- plot_list[[length(plot_list)]] +
        ggplot2::theme(axis.text.x=ggplot2::element_text(angle = 45, size = 9), axis.ticks.x = ggplot2::element_line())

      # change the y-axis tick to only max value
      ymaxs<- purrr::map_dbl(plot_list, extract_max)
      plot_list<- purrr::map2(plot_list, ymaxs, function(x,y) x +
                                ggplot2::scale_y_continuous(breaks = c(y)) +
                                ggplot2::expand_limits(y = y))

      p<- patchwork::wrap_plots(plotlist = plot_list, ncol = 1)
      return(suppressWarnings(p))
    }

    IL_topgenes <- c("REG1A","REG1B", "DMBT1" )
    IL_topgenes_plot <- suppressWarnings(StackedVlnPlot(obj = object, features = IL_topgenes, split.by = "condition", cols = condition.cols, group.by = "Final.IDS") +
      ggplot2::theme(panel.grid.major = ggplot2::element_line(colour = "gray",size=0.75)))



    if (save == TRUE){
      ggplot2::ggsave(plot=IL_topgenes_plot, file= "/Figure_S7D.pdf", width = 12, height = 7.2)
    }
    return(suppressWarnings(IL_topgenes_plot))
  }

}



#' Plot trajectory figures
#'
#' Reproduces figure panels related to differentiation trajectory modelling
#'
#' @details This function reproduces the following figure panels: \itemize{\item
#'   \strong{Figure S3B}: DiffusionMaps projection of the stem cell trajectory,
#'   colored by cell-type cluster and cell-cycle phase \item \strong{Figure
#'   S3C}: DiffusionMaps projections of the enterocyte, secretory and Paneth
#'   cell trajectories, colored by cell-type cluster}
#'
#' @param seurat.obj The list of Seurat objects (per trajectory) generated by
#'   \code{\link{model_diffusion}}
#' @param slingshot.obj The list of slingshot objects (per trajectory) generated
#'   by \code{\link{model_slingshot}}
#' @param figure ID of the figure to be reproduced. Accepted values are "S3B"
#'   and "S3C" (See Details for description of figure panels)
#' @param save Save the specified plot as a pdf? (default = FALSE)
#' @param save.dir Path to the directory where the output plot should be saved
#'   (default = ".")
#'
#' @return The plot specified in the "figure" argument
#' @export
#'
#' @examples
#' \dontrun{
#' # Make each plot and save as pdf
#' plot_trajectory(seurat.obj = diffusion.output, slingshot.obj = slingshot.output,
#'                 figure = "S3B", save = T, save.dir = ".")
#' plot_trajectory(seurat.obj = diffusion.output, slingshot.obj = slingshot.output,
#'                 figure = "S3C", save = T, save.dir = ".")}

plot_trajectory <- function(seurat.obj = NULL, slingshot.obj = NULL, figure = NA, save=FALSE, save.dir="."){
  # Check inputs
  accept.figs <- c("S3B","S3C")

  if (is.na(figure)){
    stop("Please specify a figure to plot")
  }

  if (!is.element(figure,accept.figs)){
    stop("Invalid figure ID, please specify one of the following: ", paste(accept.figs,collapse = ", "))
  }

  # General variables

  clust.order <- c("ISC","TA1","TA2","Early EC","EC","Secretory progenitor","Goblet","Paneth","EEC","Tuft")
  clust.cols <- c("#8b8c89","#FAB198", "#BF212F","#33AD75","#006F3C","#AACEDC","#254B95","#FFB60A","#CB157C","#76018D")

  # Plot Figure S3B (STEM trajectory, cluster + Phase)
  if (figure == "S3B"){

    data <- Seurat::FetchData(seurat.obj$stem, vars = c("DC_1","DC_2","Final.IDS","condition","Phase"),
                              cells = Seurat::WhichCells(seurat.obj$stem, expression = condition == "IL22+"))
    data$Phase <- factor(data$Phase, levels=c("G1","S","G2M"))

    pl1<- ggplot2::ggplot(data, ggplot2::aes(x=DC_1, y=DC_2, color=Final.IDS)) +
      ggplot2::geom_point(shape=20, size=1.5) +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
            panel.grid.minor = ggplot2::element_blank(),
            panel.background = ggplot2::element_rect(colour = "black", size=1),
            axis.ticks.length = ggplot2::unit(.25, "cm"),
            axis.title = ggplot2::element_text(size=14, face="bold"),
            axis.text = ggplot2::element_text(size=12),
            legend.title=ggplot2::element_text(size=14),
            legend.text = ggplot2::element_text(size=12),
            legend.position = c(.15, .15)) +
      ggplot2::scale_color_manual(values = clust.cols,
                         breaks = clust.order) +
      ggplot2::labs(color="Cluster") +
      ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size = 6))) +
      ggplot2::xlab("DC 1") +
      ggplot2::ylab("DC 2")

    pl2 <- ggplot2::ggplot(data, ggplot2::aes(x=DC_1, y=DC_2, color=Phase)) +
      ggplot2::geom_point(shape=20, size=1.5) +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
            panel.grid.minor = ggplot2::element_blank(),
            panel.background = ggplot2::element_rect(colour = "black", size=1),
            axis.ticks.length = ggplot2::unit(.25, "cm"),
            axis.title = ggplot2::element_text(size=14, face="bold"),
            axis.text = ggplot2::element_text(size=12),
            legend.title=ggplot2::element_text(size=14),
            legend.text = ggplot2::element_text(size=12),
            legend.position = c(.15, .15)) +
      ggplot2::labs(color="Phase") +
      ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size = 6))) +
      ggplot2::xlab("DC 1") +
      ggplot2::ylab("DC 2")

    plot <- cowplot::plot_grid(pl1,pl2)


    if (save == TRUE){
      ggplot2::ggsave(plot=plot, file = paste0(save.dir,"/Figure_S3B.pdf"), width = 12, height = 6)
    }

    return(plot)
  }

  # Plot Figure S3C (all other trajectories)

  if (figure == "S3C"){

    data <- lapply(seurat.obj, Seurat::FetchData,  vars = c("DC_1","DC_2","Final.IDS","condition"))
    data <- lapply(data, function(x) dplyr::filter(x, condition == "IL22+"))

    # Enterocyte

    plot.ec <- ggplot2::ggplot(data$ec, ggplot2::aes(x=DC_1, y=DC_2)) +
      ggplot2::geom_point(ggplot2::aes(color=Final.IDS),shape=20, size=1.5) +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
                     panel.grid.minor = ggplot2::element_blank(),
                     panel.background = ggplot2::element_rect(colour = "black", size=1),
                     axis.ticks.length = ggplot2::unit(.25, "cm"),
                     axis.title = ggplot2::element_text(size=14, face="bold"),
                     axis.text = ggplot2::element_text(size=12),
                     legend.title=ggplot2::element_text(size=14),
                     legend.text = ggplot2::element_text(size=12),
                     legend.position = c(.75, .2)) +
      ggplot2::scale_color_manual(values = clust.cols,
                                  breaks = clust.order) +
      ggplot2::labs(color="Cluster") +
      ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size = 6))) +
      ggplot2::xlab("DC 1") +
      ggplot2::ylab("DC 2") +
      ggplot2::geom_line(data=slingshot::slingCurves(slingshot.obj$ec,as.df=T),
                         ggplot2::aes(x=DC_1,y=DC_2,group=Lineage),color="grey")

    # Secretory

    plot.sec <- ggplot2::ggplot(data$secretory, ggplot2::aes(x=DC_1, y=DC_2)) +
      ggplot2::geom_point(ggplot2::aes(color=Final.IDS),shape=20, size=1.5) +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
            panel.grid.minor = ggplot2::element_blank(),
            panel.background = ggplot2::element_rect(colour = "black", size=1),
            axis.ticks.length = ggplot2::unit(.25, "cm"),
            axis.title = ggplot2::element_text(size=14, face="bold"),
            axis.text = ggplot2::element_text(size=12),
            legend.title=ggplot2::element_text(size=14),
            legend.text = ggplot2::element_text(size=12),
            legend.position = c(.75, .25)) +
      ggplot2::scale_color_manual(values = clust.cols,
                         breaks = clust.order) +
      ggplot2::labs(color="Cluster") +
      ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size = 6))) +
      ggplot2::xlab("DC 1") +
      ggplot2::ylab("DC 2") +
      ggplot2::geom_line(data=slingshot::slingCurves(slingshot.obj$secretory,as.df=T),
                         ggplot2::aes(x=DC_1,y=DC_2,group=Lineage),color="grey")


    # Paneth

    plot.paneth <- ggplot2::ggplot(data$paneth, ggplot2::aes(x=DC_1, y=DC_2)) +
      ggplot2::geom_point(ggplot2::aes(color=Final.IDS),shape=20, size=1.5) +
      ggplot2::theme_bw() +
      ggplot2::theme(panel.grid.major = ggplot2::element_blank(),
                     panel.grid.minor = ggplot2::element_blank(),
                     panel.background = ggplot2::element_rect(colour = "black", size=1),
                     axis.ticks.length = ggplot2::unit(.25, "cm"),
                     axis.title = ggplot2::element_text(size=14, face="bold"),
                     axis.text = ggplot2::element_text(size=12),
                     legend.title=ggplot2::element_text(size=14),
                     legend.text = ggplot2::element_text(size=12),
                     legend.position = c(.25, .2)) +
      ggplot2::scale_color_manual(values = clust.cols,
                                  breaks = clust.order) +
      ggplot2::labs(color="Cluster") +
      ggplot2::guides(color = ggplot2::guide_legend(override.aes = list(size = 6))) +
      ggplot2::xlab("DC 1") +
      ggplot2::ylab("DC 2") +
      ggplot2::geom_line(data=slingshot::slingCurves(slingshot.obj$paneth,as.df=T),
                         ggplot2::aes(x=DC_1,y=DC_2,group=Lineage),color="grey")

    full.plot <- cowplot::plot_grid(plotlist = list(plot.ec,plot.sec,plot.paneth), nrow = 1)

    if (save == TRUE){
      ggplot2::ggsave(plot=full.plot, file = paste0(save.dir,"/Figure_S3C.pdf"), width = 16, height = 5)
    }

    return(full.plot)
    }


}
